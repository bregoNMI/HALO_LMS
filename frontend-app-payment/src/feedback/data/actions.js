export const ADD_MESSAGE = 'ADD_MESSAGE';
export const REMOVE_MESSAGE = 'REMOVE_MESSAGE';
export const CLEAR_MESSAGES = 'CLEAR_MESSAGES';

export const addMessage = (code, userMessage, data, messageType, fieldName = null) => ({
  type: ADD_MESSAGE,
  payload: {
    code,
    userMessage,
    data,
    messageType,
    fieldName,
  },
});

export const removeMessage = id => ({
  type: REMOVE_MESSAGE,
  payload: {
    id,
  },
});

export const clearMessages = () => ({
  type: CLEAR_MESSAGES,
});
