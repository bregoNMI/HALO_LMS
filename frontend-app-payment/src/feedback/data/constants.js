export const ALERT_TYPES = {
  DANGER: 'danger',
  SUCCESS: 'success',
  INFO: 'info',
  WARNING: 'warning',
};

export const MESSAGE_TYPES = {
  DEBUG: 'debug',
  INFO: 'info',
  SUCCESS: 'success',
  WARNING: 'warning',
  ERROR: 'error',
};
