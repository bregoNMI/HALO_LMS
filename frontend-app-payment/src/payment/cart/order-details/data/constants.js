// Product types
export const COURSE_ENTITLEMENT_PRODUCT_TYPE = 'Course Entitlement';
export const ENROLLMENT_CODE_PRODUCT_TYPE = 'Enrollment Code';
export const SEAT_PRODUCT_TYPE = 'Seat';

// Certificate types
export const VERIFIED_CERTIFICATE_TYPE = 'verified';
export const CREDIT_CERTIFICATE_TYPE = 'credit';

// Order Details message keys
export const ENTITLEMENT_MESSAGE_KEY = 'entitlement';
export const ENROLLMENT_CODE_MESSAGE_KEY = 'enrollment.code';
export const VERIFIED_SEAT_MESSAGE_KEY = 'seat.verified';
export const CREDIT_SEAT_MESSAGE_KEY = 'seat.credit';
export const SEAT_MESSAGE_KEY = 'seat';
