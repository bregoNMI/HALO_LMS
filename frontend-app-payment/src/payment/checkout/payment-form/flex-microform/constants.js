export const STATUS_LOADING = 'LOADING';
export const STATUS_READY = 'READY';
export const DEFAULT_STATUS = STATUS_LOADING;
