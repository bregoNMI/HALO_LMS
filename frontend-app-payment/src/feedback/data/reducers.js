import without from 'lodash.without';

import { ADD_MESSAGE, REMOVE_MESSAGE, CLEAR_MESSAGES } from './actions';

export const storeName = 'feedback';

const defaultState = {
  byId: {},
  orderedIds: [],
};

let lastId = 0;

function getNextId() {
  return lastId++; // eslint-disable-line no-plusplus
}

export function resetIds() {
  lastId = 0;
}

const addMessage = (state, action) => {
  const id = getNextId();
  return {
    byId: {
      ...state.byId,
      [id]: {
        id,
        ...action.payload, // code, userMessage, data, messageType, fieldName
      },
    },
    orderedIds: [...state.orderedIds, id],
  };
};

const removeMessage = (state, action) => {
  const byId = { ...state.byId };
  delete byId[action.payload.id];
  return {
    byId,
    orderedIds: without([...state.orderedIds], action.payload.id),
  };
};

const reducer = (state = defaultState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case ADD_MESSAGE:
        return addMessage(state, action);
      case REMOVE_MESSAGE:
        return removeMessage(state, action);
      case CLEAR_MESSAGES:
        return defaultState;
      default:
    }
  }
  return state;
};

export default reducer;
