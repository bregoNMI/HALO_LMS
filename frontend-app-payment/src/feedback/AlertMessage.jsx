import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { Alert } from '@openedx/paragon';
import { ALERT_TYPES, MESSAGE_TYPES } from './data/constants';

// Put in a message type, get an alert type.
const severityMap = {
  [MESSAGE_TYPES.DEBUG]: ALERT_TYPES.WARNING,
  [MESSAGE_TYPES.INFO]: ALERT_TYPES.INFO,
  [MESSAGE_TYPES.SUCCESS]: ALERT_TYPES.SUCCESS,
  [MESSAGE_TYPES.WARNING]: ALERT_TYPES.WARNING,
  [MESSAGE_TYPES.ERROR]: ALERT_TYPES.DANGER,
};

const AlertMessage = (props) => {
  const {
    id, messageType, userMessage, closeHandler, data,
  } = props;

  const statusAlertProps = {
    variant: ALERT_TYPES.WARNING,
    onClose: useCallback(() => { closeHandler(id); }, [closeHandler, id]),
    show: true,
  };

  if (messageType !== null && severityMap[messageType] !== undefined) {
    statusAlertProps.variant = severityMap[messageType];
  }

  // The user message can be a
  // - React component definition (we must create it with props)
  // - React element instance (we must clone it with props)
  // - A string or number (we will render this as is)
  if (typeof userMessage === 'function') {
    statusAlertProps.dialog = React.createElement(userMessage, { values: data });
  } else if (React.isValidElement(userMessage)) {
    statusAlertProps.dialog = React.cloneElement(userMessage, {
      ...userMessage.props,
      values: { ...data, ...userMessage.props.values },
    });
  } else {
    statusAlertProps.dialog = userMessage;
  }

  return (
    <Alert {...statusAlertProps} dismissible>
      {statusAlertProps.dialog}
    </Alert>
  );
};

AlertMessage.propTypes = {
  id: PropTypes.number.isRequired,
  messageType: PropTypes.string,
  userMessage: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  closeHandler: PropTypes.func.isRequired,
  data: PropTypes.object, // eslint-disable-line react/forbid-prop-types
};

AlertMessage.defaultProps = {
  userMessage: null,
  messageType: undefined,
  data: {},
};

export default AlertMessage;
