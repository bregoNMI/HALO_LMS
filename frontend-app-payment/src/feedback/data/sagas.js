import { put } from 'redux-saga/effects';
import { getConfig, ensureConfig } from '@edx/frontend-platform';

import { logError, logInfo } from '@edx/frontend-platform/logging';
import { addMessage, clearMessages } from './actions';
import { MESSAGE_TYPES } from './constants';

ensureConfig(['ECOMMERCE_BASE_URL'], 'Alert saga');

export function* handleErrors(e, clearExistingMessages) {
  if (clearExistingMessages) {
    yield put(clearMessages());
  }

  // If this doesn't contain anything we understand, add a fallback error message
  if (e.errors === undefined && e.fieldErrors === undefined && e.messages === undefined) {
    yield put(addMessage('fallback-error', null, {}, MESSAGE_TYPES.ERROR));
  }
  if (e.errors !== undefined) {
    for (let i = 0; i < e.errors.length; i++) { // eslint-disable-line no-plusplus
      const error = e.errors[i];
      if (error.code === 'basket-changed-error-message') {
        yield put(addMessage(error.code, error.userMessage, {}, MESSAGE_TYPES.ERROR));
      } else if (error.data === undefined && error.messageType === null) {
        yield put(addMessage('transaction-declined-message', error.userMessage, {}, MESSAGE_TYPES.ERROR));
      } else {
        yield put(addMessage(error.code, error.userMessage, error.data, error.messageType));
      }
    }
  }
  if (e.messages !== undefined) {
    for (let i = 0; i < e.messages.length; i++) { // eslint-disable-line no-plusplus
      const message = e.messages[i];
      yield put(addMessage(message.code, message.userMessage, message.data, message.messageType));
    }
  }
}

export function* handleMessages(messages, clearExistingMessages, url) {
  // If this doesn't contain anything we understand, bail.
  if (!Array.isArray(messages)) {
    return null;
  }

  if (clearExistingMessages) {
    yield put(clearMessages());
  }

  // Display an error message if one is passed through as a url parameter
  if (url && url.indexOf('error_message') > 0) {
    const errorMessage = new URLSearchParams(url).get('error_message');
    yield put(addMessage('error_message', `${errorMessage}`, {}, MESSAGE_TYPES.ERROR));
  }

  for (let i = 0; i < messages.length; i++) { // eslint-disable-line no-plusplus
    const message = messages[i];
    yield put(addMessage(message.code, message.userMessage, message.data, message.messageType));
  }
  return null;
}

/**
 * handleSDNCheckFailure
 * log the SDN check failure error message
 * and redirect user to SDN failure page
*/
const handleSDNCheckFailure = (error) => {
  const setLocation = href => { global.location.href = href; };
  logError(error, {
    messagePrefix: 'SDN Check Error',
    paymentMethod: 'Stripe',
    paymentErrorType: 'SDN Check Submit Api',
    program_uuid: error?.data?.program_uuid,
  });
  if (getConfig().ENVIRONMENT !== 'test') {
    // SDN failure: redirect to Ecommerce SDN error page.
    setLocation(`${getConfig().ECOMMERCE_BASE_URL}/payment/sdn/failure/`);
  }
};

/**
 * Handle Subscription Errors
 */
export function* handleSubscriptionErrors(e, clearExistingMessages) {
  if (clearExistingMessages) {
    yield put(clearMessages());
  }
  // If this doesn't contain anything we understand, add a fallback error message
  if (e.errors === undefined && e.fieldErrors === undefined && e.messages === undefined) {
    yield put(addMessage('fallback-error', null, {}, MESSAGE_TYPES.ERROR));
  }
  if (e.errors !== undefined) {
    for (let i = 0; i < e.errors.length; i++) { // eslint-disable-line no-plusplus
      const error = e.errors[i];
      if (error.code === 'sdn_check_failure') {
        handleSDNCheckFailure(error);
      } else if (error.code === 'empty_subscription') {
        // do nothing as empty placeholder message will be rendered by Subscription Page
      } else {
        const customErrors = [
          'embargo_error',
          'basket_changed_error',
          'program_unavailable',
          'ineligible_program',
          'payment_attachment_error',
          'requires_payment_method',
        ];

        if (customErrors.includes(error.code)) {
          if (error.code !== 'create-paymentMethod') { // already logged error
            logInfo('API Error', error.code);
          }
          yield put(addMessage(error.code, error.userMessage, error?.data, MESSAGE_TYPES.ERROR));
        } else {
          logError(error.code, {
            userMessage: error.userMessage,
            errorCode: error.code,
          });
          yield put(addMessage('fallback-error', error.userMessage, error?.data, MESSAGE_TYPES.ERROR));
        }
      }
    }
  }
}
