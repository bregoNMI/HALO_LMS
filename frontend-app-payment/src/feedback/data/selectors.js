import { createSelector } from 'reselect';

import { storeName } from './reducers';

const messagesSelector = state => state[storeName];

export const messageListSelector = createSelector(
  messagesSelector,
  messages => messages.orderedIds.map(id => messages.byId[id]),
);

export const alertListMapStateToProps = createSelector(
  messageListSelector,
  messageList => ({
    messageList,
  }),
);

const messagesAsArraySelector = createSelector(
  messagesSelector,
  messages => Object.values(messages.byId),
);

const fieldNameSelector = (state, props) => props.name;

/**
 * In order to use this selector, a component must have a string "name" prop that tells the selector
 * which field to return the messages for.
 */
export const fieldMessagesSelector = createSelector(
  fieldNameSelector,
  messagesAsArraySelector,
  (fieldName, messages) => {
    const fieldMessages = [];
    messages.forEach((message) => {
      if (message.fieldName === fieldName) {
        fieldMessages.push(message);
      }
    });
    return fieldMessages;
  },
);
