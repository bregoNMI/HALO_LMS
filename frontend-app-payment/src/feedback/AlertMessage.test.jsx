import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { IntlProvider } from '@edx/frontend-platform/i18n';

import AlertMessage from './AlertMessage';
import { MESSAGE_TYPES } from './data/constants';

describe('AlertMessage', () => {
  // The AlertList test covers most of AlertMessage testing.

  it('should handle closing', () => {
    const closeHandlerMock = jest.fn();

    const component = (
      <IntlProvider locale="en">
        <AlertMessage
          id={123}
          messageType={MESSAGE_TYPES.ERROR}
          userMessage="Wondrous message!"
          closeHandler={closeHandlerMock}
        />
      </IntlProvider>
    );

    const { container } = render(component);
    const button = container.querySelector('.btn');
    fireEvent.click(button);

    expect(closeHandlerMock).toHaveBeenCalledWith(123);
  });

  it('should default its severity when necessary', () => {
    const closeHandlerMock = jest.fn();

    const component = (
      <IntlProvider locale="en">
        <AlertMessage
          id={123}
          messageType="unknown"
          userMessage="Wondrous message!"
          closeHandler={closeHandlerMock}
        />
      </IntlProvider>
    );

    const { container: tree } = render(component);
    // The alert should have an 'alert-warning' class.  That's the default in the code.
    expect(tree).toMatchSnapshot();
  });

  it('should render a userMessage function', () => {
    const component = (
      <IntlProvider locale="en">
        <AlertMessage
          id={123}
          userMessage={() => 'Wondrous message!'}
          closeHandler={jest.fn()}
        />
      </IntlProvider>
    );

    const { container: tree } = render(component);
    expect(tree).toMatchSnapshot();
  });

  it('should render a userMessage element', () => {
    const component = (
      <IntlProvider locale="en">
        <AlertMessage
          id={123}
          userMessage={<span>Wondrous message!</span>}
          closeHandler={jest.fn()}
        />
      </IntlProvider>
    );

    const { container: tree } = render(component);
    expect(tree).toMatchSnapshot();
  });
});
