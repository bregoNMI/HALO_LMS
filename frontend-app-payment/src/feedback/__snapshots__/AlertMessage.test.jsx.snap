// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertMessage should default its severity when necessary 1`] = `
<div>
  <div
    class="fade alert-content alert alert-warning show"
    dialog="Wondrous message!"
    role="alert"
  >
    <div
      class="pgn__alert-message-wrapper"
    >
      <div
        class="alert-message-content"
      >
        Wondrous message!
      </div>
      <div
        class="pgn__alert-actions pgn__action-row"
      >
        <span
          class="pgn__action-row-spacer"
        />
        <button
          class="btn btn-tertiary btn-sm"
          type="button"
        >
          Dismiss
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`AlertMessage should render a userMessage element 1`] = `
<div>
  <div
    class="fade alert-content alert alert-warning show"
    dialog="[object Object]"
    role="alert"
  >
    <div
      class="pgn__alert-message-wrapper"
    >
      <div
        class="alert-message-content"
      >
        <span
          values="[object Object]"
        >
          Wondrous message!
        </span>
      </div>
      <div
        class="pgn__alert-actions pgn__action-row"
      >
        <span
          class="pgn__action-row-spacer"
        />
        <button
          class="btn btn-tertiary btn-sm"
          type="button"
        >
          Dismiss
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`AlertMessage should render a userMessage function 1`] = `
<div>
  <div
    class="fade alert-content alert alert-warning show"
    dialog="[object Object]"
    role="alert"
  >
    <div
      class="pgn__alert-message-wrapper"
    >
      <div
        class="alert-message-content"
      >
        Wondrous message!
      </div>
      <div
        class="pgn__alert-actions pgn__action-row"
      >
        <span
          class="pgn__action-row-spacer"
        />
        <button
          class="btn btn-tertiary btn-sm"
          type="button"
        >
          Dismiss
        </button>
      </div>
    </div>
  </div>
</div>
`;
